Improve unit tests performance (CASSANDRA-17427 --> trunk)

jacek-lewandowski opened the pull request:
- Don't flush schema on every schema update in unit tests
- Use unix command to delete test data
- Shorten teardown
- Stable processor count presented by JMX on Jenkins, CircleCI and local

patch by @jacek-lewandowski reviewed by @michaelsembwever and @josh-mckenzie for [CASSANDRA-17427](https://issues.apache.org/jira/browse/CASSANDRA-17427)

michaelsembwever commented on test/unit/org/apache/cassandra/tools/ToolRunner.java:
```diff
@@ -216,6 +216,8 @@ public static ObservableTool invokeAsync(Map<String, String> env, InputStream st
         ProcessBuilder pb = new ProcessBuilder(args);
         if (env != null && !env.isEmpty())
             pb.environment().putAll(env);
+        String jvmOpts = pb.environment().getOrDefault("JVM_OPTS", "") + " -Dcassandra.disable_tcactive_openssl=true";
```
why also put `-Dcassandra.disable_tcactive_openssl=true` here, as well as into the build.xml ?
 e.g. see https://github.com/apache/cassandra/pull/1488/files#diff-766797f233c18114f9499750cf1ffbf3829aeea50283850619c01bd173132021R1488

jacek-lewandowski replied:
well, because we want the subprocess also use that property - it is not inherited automatically, right?

michaelsembwever commented on src/java/org/apache/cassandra/io/util/PathUtils.java:
```diff
@@ -63,14 +63,16 @@
     private static final Set<StandardOpenOption> READ_WRITE_OPTIONS = unmodifiableSet(EnumSet.of(READ, WRITE, CREATE));
     private static final FileAttribute<?>[] NO_ATTRIBUTES = new FileAttribute[0];
 
+    private static final boolean USE_NIX_RECURSIVE_DELETE = CassandraRelevantProperties.USE_NIX_RECURSIVE_DELETE.getBoolean();
```
Could this condition be automatic on *nix, instead of explicit via a property flag?

Is this applicable to runtime as well as tests?

jacek-lewandowski replied:
I wanted to make this a simple fix for testings - I guess changing that in general would require some extended verification

michaelsembwever replied:
fair enough. maybe a comment that we eventually want to hard code it to true? 
(and how we will encourage testing/verification of it?)

jmckenzie-dev replied:
Why alias this locally inside PathUtils.java instead of just calling the ``CassandraRelevantProperties.USE_NIX_RECURSIVE_DELETE.getBoolean()`` on each access?

jacek-lewandowski replied:
Looking into other classes I perceived it as a convention, but I'm not opposed to change that

jmckenzie-dev replied:
Eh, either way is fine. Just curious.

michaelsembwever commented on build.xml:
```diff
@@ -888,8 +888,7 @@
     -->
-    <target name="build" depends="resolver-retrieve-build,build-project,checkstyle" description="Compile Cassandra classes"/>
-    <target name="_build_unsafe" depends="resolver-retrieve-build,build-project" description="Compile Cassandra classes without checks"/>
+    <target name="build" depends="resolver-retrieve-build,build-project" description="Compile Cassandra classes"/>
```
a concern i have is that by removing `checkstyle` from `build`, `rat-check` from `build-project`, and `eclipse-warnings` from `jar`, we are less likely to be catching these failures before commits are pushed. because folk are not often running `artifacts` locally. even if CI catches it, it is still nice to have a faster dev feedback loop on such failures. also worth noting that circleci does not run artifacts.

so the `_build_unsafe` and `*.skip` approach has appeal to me. leaning me towards instead adding the appropriate skip flags to all the ant calls in https://github.com/apache/cassandra-builds/blob/trunk/build-scripts/cassandra-test.sh and https://github.com/apache/cassandra-builds/blob/trunk/build-scripts/cassandra-dtest-pytest.sh

jacek-lewandowski replied:
Actually, they are explicitly added to `jar`: https://github.com/apache/cassandra/pull/1488/commits/04bc756409b99a34078f68c271df1f913098385b#diff-766797f233c18114f9499750cf1ffbf3829aeea50283850619c01bd173132021R1179

So the check will surely run in beginning of CI build

michaelsembwever commented:
CI ci-cassandra: https://ci-cassandra.apache.org/job/Cassandra-devbranch/1834/

jmckenzie-dev commented on ide/idea/workspace.xml:
```diff
@@ -353,4 +353,4 @@
     </entry>
   </component>
-</project>
+</project>
```
Why drop whitespace at end of file?

jmckenzie-dev commented on src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:
```diff
@@ -577,5 +581,4 @@ public boolean isPresent()
     }
-}
-
+}
```
Why the drop of whitespace at end of file?

jmckenzie-dev commented on src/java/org/apache/cassandra/io/util/PathUtils.java:
```diff
@@ -64,14 +64,16 @@
             setDeletionListener(ignore -> {});
-        else
-            logger.info("Deleting file during startup: {}", path);
+        else if (logger.isTraceEnabled())
```
Not for this patch, but we should create a follow up ticket: we should wrap this in some kind of util method instead of having this if / log at trace pattern peppered throughout the codebase. Something like `FBUtilities.logIfTrace(Logger logger, string msg)`.

jacek-lewandowski replied:
Obviously this is not needed in this case.

However, in general case we should use `if` when:
```java
log.trace("message {}", someHeavyToString());
```

if we go with a method like `logIfTrace` we would have to either evaluate that `someHeavyToString()` before going to that method regardless trace is enable or not (which undermine the existence of `if` at all) or we would have to pass a lambda, which is also much heavier than just checking `logger.isTraceEnabled`.

I think the rules should be clear there - unless we do explicit string concatenation (which enforces calling `toString()` for each element) or we call something more than just trivial getters, we should have `if`, otherwise we can skip it because it brings nothing.

On the other hand, `Logger` has methods with signatures:
```java
trace(String msg, Object ... paramas);
trace(String msg, Throwable t);
```
so basically if we want to pass a throwable we cannot make use of passing string format parameters as objects without the need to early call their `toString()` and actually doing manual string concatenation. 

in particular, there is lack of a method like:
```java
trace(String msg, Throwable t, Object ... params)
```

Such a method could be useful in that context.

jmckenzie-dev replied:
Really good point, and it'd split the idiom if we had "string interpolation goes this route, non goes that route". I've spent too much time working in C# w/Conditional runtime method attributes. :)

Could _probably_ rely on hotspot to realize don't have tracing enabled and optimize that path away but not worth the UX / idiom split IMO.

jmckenzie-dev commented on src/java/org/apache/cassandra/io/util/PathUtils.java:
```diff
@@ -310,13 +312,39 @@ public static Throwable delete(Path file, Throwable accumulate, @Nullable RateLi
+    {
+        try
+        {
+            int result = Runtime.getRuntime().exec(new String[]{ "rm", quietly ? "-rf" : "-r", path.toAbsolutePath().toString() }).waitFor();
```
nit: build the command and cache in a string locally so you eliminate the potential defect / error of us logging a command that differs from the one we ran.

jmckenzie-dev commented on src/java/org/apache/cassandra/io/util/PathUtils.java:
```diff
@@ -310,13 +312,39 @@ public static Throwable delete(Path file, Throwable accumulate, @Nullable RateLi
         return accumulate;
     }
 
+    private static void deleteRecursiveUsingNixCommand(Path path, boolean quietly)
```
May want to javadoc why this method exists and why we're using it for future maintainers. Also nit: consider renaming to `deleteRecursiveFast` which would leave us flexibility to switch on platform for faster deletes in the future if (for some bizarre reason) the rm approach ends up slower. Thinking OSX, WSL v2, etc.

jacek-lewandowski replied:
We always want to that fast don't we? 😄 

But this is a private method named after what it really does. And then we have a public remove method which decides which way to delete files should be chosen.

jmckenzie-dev replied:
Well, fair, nobody wants delete to be _slow_. But the reason this method exists is because on a specific platform it's significantly faster, and the only way a future maintainer would know that is an anthropological dig w/JIRA comments and descriptions to determine why something was changed. I'm fine leaving the name as is, but commenting why this seemingly redundant functionality was introduced would be nice.

jmckenzie-dev commented on src/java/org/apache/cassandra/net/MessagingService.java:
```diff
@@ -233,6 +234,8 @@ public static int getVersionOrdinal(int version)
         return ordinal;
     }
 
+    public final static boolean NON_GRACEFUL_SHUTDOWN = Boolean.getBoolean("cassandra.test.messagingService.nonGracefulShutdown");
```
Should this be a `CassandraRelevantProperties`?

smiklosovic replied:
@josh-mckenzie great catch, yes, it definitely should be there. If we managed to have the second reviewer for CASSANDRA-17797 and we merged that work, this would fail the build :)

jacek-lewandowski replied:
Quite interesting question. This is something to be used only in tests. Perhaps something overly contrived but maybe test properties should require something like `cassandra.test` property to be set in order to make them get assigned with a non-default value? (I mean, set `cassandra.test` to unlock those properties). Perhaps something for CASSANDRA-17797 @smiklosovic

smiklosovic replied:
The patch we work on is gathering all properties, irrelevant what they are for, in one place. I understand these are just test properties but having them somewhere separately, for example, in test source, would mean that we would start to reference test class in the prod code. Right? Not good. Having test code referencing prod code where these test properties are is better.

I am not sure how to reply to your idea about `cassandra.test`. I think that in general it is good idea to have them with that prefix, yes. Otherwise we are not completely sure, at the first glance, if there is not any comment, what that property is actually used for / upon.

In the long run, we would like to have all test properties with such prefix (it is really messy right now) but the first task is to manage the current "chaos" with having them scattered across all the source code. As of now, nobody has a definitive answer what all properties we use ....

I think test properties with some value should be set as the very first thing in the test class?  So the defaults would be overridden (if desirable) by a tester upon starting that test when this is `final static`. If the value of that property is meant to be changed during the test then this needs to be fixed on the code level to resolve that property dynamically as it is called.

jacek-lewandowski replied:
I just meant to prevent setting those properties to some non-default values in production, that is, you can only override the defaults if some additional property like, `-Dcassandra.test=true` is set

smiklosovic replied:
ah right ...  @Mmuzaf this is an interesting point to take into account, we might deal with this later I guess.

jmckenzie-dev commented on src/java/org/apache/cassandra/net/MessagingService.java:
```diff
@@ -527,11 +530,17 @@ public Future<Void> maybeReconnectWithNewIp(InetAddressAndPort address, InetAddr
     }
 
-    public void shutdown(long timeout, TimeUnit units, boolean shutdownGracefully, boolean shutdownExecutors)
+    public void shutdown(long timeout, TimeUnit units, boolean shutdownGracefully, boolean shutdownExecutors, boolean blocking)
```
I think this would be cleaner to break into 2 methods, `shutdownGracefully()` and `shutdownAbruptly()` or something, rather than going the multi-bool precondition most of the method is dead on various branches approach we have here.

jacek-lewandowski replied:
Yes, I agree

jmckenzie-dev commented on test/distributed/org/apache/cassandra/distributed/impl/AbstractCluster.java:
```diff
@@ -184,6 +184,12 @@
         private INodeProvisionStrategy.Strategy nodeProvisionStrategy = INodeProvisionStrategy.Strategy.MultipleNetworkInterfaces;
         private ShutdownExecutor shutdownExecutor = DEFAULT_SHUTDOWN_EXECUTOR;
 
+        {
```
Is there a reason you went w/the instance initializer vs. static initializer? I'd expect the latter to appropriately nuke the param for everything created inside the JVM context but I could be missing something.

jacek-lewandowski replied:
No particular reason AFAIR but the cluster initialization seemed more consistent to me that a static initializer because those properties are really relevant for the instance rather than static class. If we ever come up with the ability to pass individual system properties to node instances in JVM dtests, setting properties in a static initializer will become not relevant a all

jmckenzie-dev reviewed (changes requested):
No really major issues on the review; looking solid.

jmckenzie-dev commented:
In general there's a lot of whitespace removal in various files; I'm not sure why we have so many files w/trailing whitespace and we don't have anything in our style guide about it that I can see ([link](https://cassandra.apache.org/_/development/code_style.html)). So is without or with the way to go?

jacek-lewandowski commented:
@josh-mckenzie I've applied your review comments, please have a look before I rebase and overwrite the branch

smiklosovic commented on src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:
```diff
@@ -584,4 +612,3 @@ public boolean isPresent()
         return System.getProperties().containsKey(key);
     }
 }
-
```
@jacek-lewandowski this should be probably undeleted.

jacek-lewandowski replied:
I'll make sure there is no such changes before committing/merging

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/impl/AbstractCluster.java:
```diff
@@ -1319,4 +1326,3 @@ private enum Utils implements NameHelper
         INSTANCE;
     }
 }
-
```
@jacek-lewandowski same

jacek-lewandowski replied:
actually - no, I'll not revert that - I've just looked and in those two files there were 2 empty lines in the end. Now there is one

smiklosovic replied:
ok thanks for the clarification

jacek-lewandowski committed:
Improve unit tests performance

- Don't flush schema on every schema update in unit tests
- Use unix command to delete test data
- Shorten teardown
- Stable processor count presented by JMX on Jenkins, CircleCI and local

Patch by <jacek-lewandowski>, reviewed by <michaelsembwever> and <josh-mckenzie> for CASSANDRA-17427

jacek-lewandowski committed:
DO NOT MERGE - CircleCI config
